core/dispatch.c

```diff
@@ -874,6 +874,17 @@ dispatch_enter_dynamorio(dcontext_t *dcontext)
         if (get_at_syscall(dcontext))
             handle_post_system_call(dcontext);
 
+#ifdef X86
+        /* If the next basic block starts at a debug register value,
+         * we fire a single step exception before getting to the basic block. */
```

Comment:
Ok, I had first misunderstood your comment.

With this patch, the client will not see the bb with this instruction until after the fault is raised and handled.

It is the same with a mid-bb debug reg : the patch forces the basic block to end before getting to a debug reg.